rs/consensus/src/consensus/block_maker.rs

```diff
@@ -514,17 +516,51 @@ pub(crate) fn already_proposed(pool: &PoolReader<'_>, h: Height, this_node: Node
         .any(|p| p.signature.signer == this_node)
 }
 
+const DYNAMIC_DELAY_LOOK_BACK_DISTANCE: Height = Height::new(29);
+const DYNAMIC_DELAY_MAX_NON_RANK_0_BLOCKS: usize = 5;
+const DYNAMIC_DELAY_EXTRA_DURATION: Duration = Duration::from_secs(10);
+
+fn count_non_rank_0_blocks(pool: &PoolReader, block: Block) -> usize {
+    let max_height = block.height();
+    let min_height = if max_height > DYNAMIC_DELAY_LOOK_BACK_DISTANCE {
+        max_height - DYNAMIC_DELAY_LOOK_BACK_DISTANCE
+    } else {
+        Height::new(0)
+    };
+
+    pool.get_range(block, min_height, max_height)
+        .filter(|block| block.rank > Rank(0))
+        .count()
+}
+
 /// Calculate the required delay for block making based on the block maker's
-/// rank.
+/// rank and the number of non-0-rank blocks in its ancestry.
 pub(super) fn get_block_maker_delay(
     log: &ReplicaLogger,
     registry_client: &dyn RegistryClient,
     subnet_id: SubnetId,
+    pool: &PoolReader<'_>,
+    parent: Block,
     registry_version: RegistryVersion,
     rank: Rank,
+    metrics: Option<&BlockMakerMetrics>,
 ) -> Option<Duration> {
-    get_notarization_delay_settings(log, registry_client, subnet_id, registry_version)
-        .map(|settings| settings.unit_delay * rank.0 as u32)
+    let settings =
+        get_notarization_delay_settings(log, registry_client, subnet_id, registry_version)?;
+    // If this is not a Rank-0 block maker, check how many non-rank-0 blocks have been notarized in
+    // the past, and increase the delay if there have been too many.
+    let dynamic_delay = if rank > Rank(0)
```

Comment:
cc: FYI @DSharifi maybe the dynamic delay that we discuss will capture the situation this MR is trying to prevent.

no action here